sdk/storage/azure-storage-datamovement/src/job_engine_task_generation.cpp

```diff
@@ -146,14 +250,28 @@ namespace Azure { namespace Storage { namespace _detail {
       // TODO: It's not a good idea to invoke a network request when generating tasks, as this
       // function is executed in single thread so it's likely to become a bottleneck. Find some way
       // to optimize this. This also applies to downloading directory, downloading page blob etc.
-      int64_t fileSize = m_model.Source.m_blobClient.Value().GetProperties().Value.BlobSize;
+      auto blobProperties = m_model.Source.m_blobClient.Value().GetProperties().Value;
+      int64_t fileSize = blobProperties.BlobSize;
 
       TaskModel task;
       task.ObjectSize = fileSize;
       task.ChunkSize = g_DownloadBlockSize;
       task.NumSubtasks
           = static_cast<int32_t>((fileSize + g_UploadBlockSize - 1) / g_DownloadBlockSize);
       task.NumSubtasks = std::max(task.NumSubtasks, 1);
+      if (blobProperties.BlobType == Blobs::Models::BlobType::BlockBlob)
+      {
+        task.SourceType = TaskModel::ResourceType::BlockBlob;
+      }
+      else if (blobProperties.BlobType == Blobs::Models::BlobType::AppendBlob)
+      {
+        task.SourceType = TaskModel::ResourceType::AppendBlob;
+      }
+      else if (blobProperties.BlobType == Blobs::Models::BlobType::PageBlob)
+      {
+        task.SourceType = TaskModel::ResourceType::PageBlob;
+        PageBlobTaskDecoration(task, m_model.Source.m_blobClient.Value().AsPageBlobClient());
```

Comment:
List page range could take pretty long time. Don't put it in the main scheduling thread.

Reply:
Good point. It's not a good idea to issue network requests in scheduling thread. This can be a bottleneck especially when downloading large number of small files/blobs. Listing page ranges, by comparison, is less of a problem. Because small files can be as small as several bytes, and each file/blob item takes hundreds of bytes in ListBlobs response. However, page range is at least 512 bytes, and a range takes only like ~20 bytes in ListPageRanges response.

Anyway, we already have a [TODO](https://github.com/Azure/azure-sdk-for-cpp/blob/dcf62d20eb42f42c40507b8ef6445a5a8730662e/sdk/storage/azure-storage-datamovement/src/job_engine_task_generation.cpp#L146) item here. We also have GitHub issue #3603 tracking these work items. The purpose of this PR is to make most common scenarios work. We'll make optimizations for other scenarios in separate PRs. I hope this all makes sense to you.